Arr integration

tachukao opened the pull request:
1. allow users to integrate Ndarrays of dimension n. the results is an Ndarray of dimension (n+1), where the first dimension corresponds to time
2. expose type `tspec` and `Solver` directly in `Owl_ode`

tachukao committed:
support Arr integration

tachukao committed:
include types tspec and Solver into owl_ode

tachukao commented:
argh...Travis is failing because the current version of `owl` on `opam` doesn't have `expand` in `owl_types_ndarray_basic.ml` yet

ryanrhymes commented:
@mseri we probably need a new release of Owl?

mseri commented:
I can draft one later today

ryanrhymes commented:
Awesome!

mseri commented:
There are some inconsistencies with the tests. I’ll probably make a release with testing disabled for now

mseri commented:
In progress: https://github.com/ocaml/opam-repository/pull/15278

mseri commented on src/ode/owl_ode.ml:
```diff
@@ -3,3 +3,4 @@ module Ode = Owl_ode_base.Ode
 module Types = Owl_ode_base.Types
 module Native = Native
 module Symplectic = Symplectic
+include Types
```
How much will this mess up the types? Types from different modules would become different right? I'd rather keep it spearate as before

tachukao replied:
yea that's true I didn't think of that. I'll revert it

mseri replied:
If you want, you could add `owl_ode.mli` and expose the types there by hand. In the implementation you would just have `val tspec = Types.tspec` while in the signature you can fully expand it (you will also have to copy-paste the docstring but I guess we can live with some level of duplication). I think that should work.

tachukao replied:
argh of course...that would be the way to do it. But perhaps, this is best left for another PR.

mseri replied:
I don’t know. It’s such a small change that we may as well add it here. Do you think it would take too long?

tachukao replied:
wouldn't we have to change the docs as well??

tachukao replied:
but yea to be fair, we'll probably have to update the docs anyway. okay, I'll make the changes expose the type `tspec`

mseri replied:
I have added a commit that solves this issue. I have also rebased on the current origin to cleanup the history from the many merges

mseri commented on src/sundials/owl_ode_sundials.ml:
```diff
@@ -76,26 +70,25 @@ let cvode_s'
       match r with
       | Success ->
-        let y' = Mat.copy (unwrap (dim1, dim2) y) in
+        let y' = Arr.copy (Arr.reshape (unwrap y) s) in
```
Can you explain me more about the rationale? Why is mixing up the types in this way, sometimes mat and sometimes arr, better?

tachukao replied:
the rationale here is to support integration of higher dimensional Ndarrays (n >= 2).  `Mat.mat` and `Arr.arr` are both `genarray` underneath.

tachukao committed:
do not include Types

tachukao committed:
fixed f_wrapped + consistently Mat.mat -> Arr.arr

mseri commented:
We should also update the lower bounds on owl and owl-base in the opam files, and the examples

tachukao committed:
updated version bounds

mseri committed:
Add interface to owl_ode to keep types transparent

It will also fail to compile if we modify the types in owl_ode but forget to change it in owl_ode_base.types and viceversa

Signed-off-by: Marcello Seri <marcello.seri@gmail.com>
